### PR TITLE
[Inference] Add llama model based on padding inputs

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
related to [#5168](https://github.com/hpcaitech/ColossalAI/issues/5168)

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/modeling/models/llama.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    # TODO: The code below will be uncommented after the development of attention-related kernel is completed.
+    # memcpy_to_block(key_states, value_states, k_cache, v_cache, block_tables, block_size, sequence_lengths)
+    # if is_prompts:
+    #     attn_output = context_attention_unpadded(query_states, key_states, value_states, k_cache, v_cache, sequence_lengths, block_tables, block_size)
+    # else:
+    #     attn_output = torch.empty(bsz, self.num_heads, self.head_dim)
+    #     decoding_attention(query_states, k_cache, v_cache, block_tables, sequence_lengths, attn_output, block_tables.shape[1], block_size)
+
+    attn_output = query_states
```
Comment (yuehuayingxueluo) [inline colossalai/inference/modeling/models/llama.py:168]: This is used for accuracy alignment with transformer.

File: colossalai/inference/core/engine.py [hunk 2]
```diff
@@ -184,13 +189,14 @@ def add_request(
                 prompt = None
             else:
                 prompt = prompts[i]
+            block_table = torch.full([self.inference_config.max_seq_len], -1, device=self.device)
```
Comment (CjhHa1) [inline colossalai/inference/core/engine.py:192]: should be `max_block_num`

File: colossalai/inference/core/request_handler.py [hunk 3]
```diff
@@ -90,17 +94,16 @@ def schedule(self):
...
                             break
                         # Try to allocate cache blocks for the sequence.
                         if self.cache_manager.check_allocation(seq):
                             # If succeed, add the sequence to running list.
                             self.running_list.append(seq)
-                            self.cache_manager.allocate_context_from_block_table(seq.block_table, seq.prompt_len)
-                            lst.remove(seq)
-
+                            self.cache_manager.allocate_context_from_block_table(seq.block_table, seq.input_len)
+                    lst.clear()
```
Comment (CjhHa1) [inline colossalai/inference/core/request_handler.py:106]: why use `clear()` here

File: colossalai/inference/modeling/models/llama.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+
+    key_states = repeat_kv(key_states, self.num_key_value_groups)
+    value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+    query_states = query_states.transpose(1, 2)
+    key_states = key_states.transpose(1, 2)
+    value_states = value_states.transpose(1, 2)
+
+    query_states = query_states.view(-1, self.num_heads, self.head_dim)
```
Comment (CjhHa1) [inline colossalai/inference/modeling/models/llama.py:156]: `transpose` and `view` can be put together

File: colossalai/inference/config.py [hunk 5]
```diff
@@ -1,7 +1,6 @@
 """
-Our config consists of two parts:
+Our config consists of one part:
     1. inference_config: configs for inference, it is a unified api that wraps all the configs for inference.
```
Comment (FrankLeeeee) [inline colossalai/inference/config.py:3]: If it is only one part, you should just phrase like `Our config contains various options for inference optimization, it is a unified API that wraps all the configurations for inference.`

File: colossalai/inference/config.py [hunk 6]
```diff
@@ -98,3 +97,6 @@ def _verify_config(self) -> None:
         ], "dtype should be one of 'fp16', 'fp32', 'bf16', torch.float32, torch.float16, torch.bfloat16"
         assert self.max_batch_size <= 64, "Max batch size exceeds the constraint"
         assert self.quant_mode in ["smoothquant", "gptq", None], "quant should be one of 'smoothquant', 'gptq'"
+        assert (
+            self.max_input_len + self.max_output_len <= self.max_seq_len
+        ), "The sum of max_input_len and max_output_len must be smaller than max_seq_len."
```
Comment (FrankLeeeee) [inline colossalai/inference/config.py:102]: For the sake of user-friendliness, you should print out the actual numbers. e.g.

```python
f"The sum of max_input_len ({self.max_input_len})  and max_output_len ({self.max_output_len}) must be smaller than max_seq_len ({self.max_seq_len})."
```

File: colossalai/inference/kv_cache/kvcache_manager.py [hunk 7]
```diff
@@ -110,6 +110,10 @@ def num_available_blocks(self) -> int:
         """Get the number of available cache blocks."""
         return self._available_blocks
 
+    def get_kv_cache(self):
+        """Get k_cache and v_cache"""
+        return self._kv_caches[0], self._kv_caches[1]
```
Comment (yuanheng-zhao) [inline colossalai/inference/kv_cache/kvcache_manager.py:115]: `self._kv_caches` itself is a tuple.

File: colossalai/inference/core/engine.py (hunk 2 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/core/engine.py:192]: We don't have a parameter called max_block_num.

Comment (CjhHa1) [reply colossalai/inference/core/engine.py:192]: https://github.com/hpcaitech/ColossalAI/blob/eb672648c89e7f0e47cb6628c7c9f044bd1137ad/colossalai/inference/kv_cache/kvcache_manager.py#L118C34-L118C34

File: colossalai/inference/config.py (hunk 5 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:3]: Done

File: colossalai/inference/config.py (hunk 6 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/config.py:102]: Done

File: colossalai/inference/core/request_handler.py (hunk 3 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/core/request_handler.py:106]: Done

File: colossalai/inference/kv_cache/kvcache_manager.py (hunk 7 above)
Comment (yuehuayingxueluo) [reply colossalai/inference/kv_cache/kvcache_manager.py:115]: Done

**Review by CjhHa1 — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**